pkg/volume/volume.go

```diff
@@ -48,6 +48,10 @@ type BlockVolume interface {
 	// and name of a symbolic link associated to a block device.
 	// ex. pods/{podUid}/{DefaultKubeletVolumeDevicesDirName}/{escapeQualifiedPluginName}/, {volumeName}
 	GetPodDeviceMapPath() (string, string)
+
+	// MetricsProvider embeds methods for exposing metrics (e.g.
+	// used, available space).
+	MetricsProvider
 }
```

Comment:
What happens for other block volume types (such as gce-pd, aws etc) where `MetricsProvider` is not initialized but now we try to grab metrics from them? Will that result in crash?

Reply:
I guess the assigning in `func (s *volumeStatCalculator) calcAndStoreStats()` would fail, as the particular blockVolume does not implement the `MetricsProvider` interface.

I'll add a type assertion there, that should be sufficient. Or do you suggest an other approach?

Reply:
We could fix other drivers to ensure they are initializing the `BlockVolume` interface correctly.  If a driver says it implements `BlockVolume` then it should implement it. Using type assertion here will fix the code but I wonder if this will cause further maintenance headaches.

Reply:
Another thing we could do is - remove `MetricsProvider` from `BlockVolume` definition here but have a type assertion like following in csi_block.go:

```
var _ volume.MetricsProvider = &csiBlockMapper{}
```
and then in `calcAndStoreStats` only pick volumes that implement `volume.MetricsProvider` . This would avoid having to initialize `MetricsProvider` in other plugins. 

Another thing though is - if we leave other plugins untouched, we will not get block volume metrics for them and hence our e2e will be kind of useless (unless it is refactored to work with csi volumes).

Reply:
I would like to implement this for in-tree drivers, similar to how basic statistics for filesystems are gathered. Unfortunately I can not tell yet when I will be able to spend time on this.